docs/design/audit-v2.md

```diff
@@ -0,0 +1,78 @@
+# Auditing V2: Random Node Selection
+
+## Abstract
+
+This design doc outlines how we will implement a Version 2 of the auditing service.
+With our current auditing service, we are currently auditing per segment of data.
+We need to keep this existing method of auditing, but also add an additional audit selection process that selects random nodes to audit.
+
+## Background
+
+As our network grows, it will take longer for nodes to get vetted.
+This is because every time an upload happens, we send 5% of the uploaded data to unvetted nodes and 95% to vetted nodes.
+When auditing occurs, we currently select a random stripe within a random segment.
+If we're selecting audits at random per byte, every segment has some percentage that went to vetted nodes.
+As more nodes join the network, it will become exponentially less likely that an unvetted node will be audited since most data will be stored on vetted nodes.
+With a satellite with one petabyte of data currently, new nodes will take one month to get vetted, which is way too long.
+
+We want a way to select segments to audit based on statistically randomly selected storage nodes.
+However, the way we select nodes should also be biased toward unvetted nodes.
+Currently, there's not a way to select a stripe based on a storage node.
+
+## Design
+
+We will create an audit observer that uses the metainfo loop, and this observer will use both methods of selection. It will create reservoir samples for segments following the current selection method (randomly across bytes), and also create a small reservoir for each unvetted node.
```

Comment:
Why not have reservoirs for all the nodes?

Reply:
We discussed this a little today, but we want to minimize memory consumption, and the primary reason we are adding this method of auditing is to give unvetted nodes more audits so they can get vetted. Once they are vetted, they should be audited as normal based on how much data they are storing, but as far as I understand, it is not super important to give already-vetted nodes more audits.

Reply:
Let's add this information to the documentation.

Reply:
maybe we need three audit loops - per segment, per node, and per unvetted node?

Reply:
when you say loops do you mean observers on the metainfo loop? we were only planning on using one observer

Reply:
i guess i should have said "audit selection strategies" instead of "audit loops"

Reply:
I don't think we need to separate "per node" and "per unvetted node". If we are going to be auditing on a per-node basis, we might as well keep them together. We keep a reservoir for every node, then we just have a query to pick some number of vetted and unvetted nodes to audit.

Reply:
or could we do what we talked about on Friday with reservoir sampling but doing it for all nodes instead of only unvetted nodes?

Reply:
Oh, I guess we could have a single loop to build the reservoirs for all the nodes as well as the "primary" segment selection reservoir, then three separate services like you say that can be configured independently.

Reply:
> maybe we need three audit loops - per segment, per node, and per unvetted node?

Aren't we overcomplicating everything?

I'd step back and ask "Why do we need auditing?". From past discussions, we agree that it is not feasible to audit all data. This is why we were selecting random stripes from random segments. So we can catch unreliable nodes and disqualify them. This randomness approach gives more pressure (i.e. audit requests) to nodes storing more data.

Now, with the above approach, we see that there are some issues. In a large network, unvetted nodes would get very few audit requests, and there are also cheater strategies that vetted nodes with a small amount of stored data could apply.

What if, we implement only a single loop with reservoir samples for all (vetted and unvetted) nodes? And ditch the old loop that select random stripes from random segments. This way we give all nodes equal chance to get audited. We will still catch if they are reliable storing data. And, if a node is failing too many audits, does it really matter how much data it stores? We would disqualify it regardless of the amount of data, right? So, sending more audit requests to nodes storing more data should not make any sense.

Reply:
We need both.
The per node selection is good to get vetted and to catch cheater. We want to detect cheater nodes within a few das.
The per byte selection is good to detect hardware failures in big nodes as quickly as possible. This time we don't want to wait a few days. The bigger the node the faster we want to detect the failure.

Reply:
Note that even with per node detection only, nodes that store more data will be audited more frequently. This is due to the fact that we don't audit just a single node at the same time (like the reverification of contained node), but we download the erasure shares from all nodes in the pointer.

The reservoir sample for each node will include pointers that for sure have a piece for this specific node. But when a stripe from this pointer is audited, the erasure shares will be downloaded from all nodes in this pointer. If a node stores more data, there is a higher chance that it has a piece from this pointer too.